README.md
shim
====

The Userify shim creates users, manages sudo permissions, etc based on
the user accounts that you've configured in the Userify web console (https://console.userify.com)
or through the API. It wakes up once every 90 seconds, checks for things to do,
and goes back to sleep.

The shim is designed to be extremely lightweight and easy to integrate and
customize into cloudinit, chef recipes, puppet manifests, RPMs, DEBs,...
We're here to help. If you experience any issues, let us know.

It has zero requirements beyond what is included in every major distribution today.


System Requirements
-------------------

The Userify shim is very short (please read it for yourself)
and only requires:

*  Linux 2.6 or later
*  curl (command-line), sudo
*  Python 2.6 or later (for httplib timeout, simplejson)
*  HTTPS access to shim.userify.com (currently proxies are not supported)

These basics are *built-in* to most Linux distributions made in the last five years,
including Red Hat, Debian, Ubuntu, RHEL, CentOS, SUSE, Gentoo,
and derivatives.

Tested distributions: RHEL7, CentOS7, Debian 7 (wheezy) and later, Ubuntu 12.04 and later.  (Free integration support available.) Currently no BSD-based UNIX (incl OS X) is supported.


What does the installer do?
---------------------------

*   Create /opt/userify and credentials file in it (creds.py)
*   Patch /etc/rc.local (/etc/init.d/after.local on SUSE)
    with a link to the daemon
*   Creates an uninstall script at /opt/userify/uninstall.sh
*   Kicks off shim every 90 seconds


Using older versions of Linux
-----------------------------

Using an older Linux version such as RHEL5 where the default Python 2.4 is nearly ten years old? You'll need to install
a Python 2.7 RPM (check DAG's repo) which will leave your existing Python 2.4 available for system usage.
Then set the PYTHON environment variable to your new version of Python in /etc/rc.local as follows:

    PYTHON=/usr/bin/python2.6 /opt/userify/shim.sh &

(Note: installer.sh will no longer start shim.sh automatically.)


Enterprise Support Available
----------------------------

For free integration support, please email support@userify.com.

To get signed up with a paid enterprise support package, please email enterprise@userify.com.


Get In Touch
------------

We are available to assist with questions, custom installations, directory
integrations or deployments, self-hosted installations, and professional
consulting. Please open an issue with your question or contact support for
assistance.




CHANGELOG
2015-09-05  Self-signed modifications to support Userify Enterprise.

2015-08-09  Major enhancements:
            *   Increased logging
            *   /opt/userify/userify_config.py
                    Optional configuration file with these options:
                    debug = False
                    dry_run = False
                    shim_host = "hostname", for enterprise use
                    self_signed = 1, supports self-signed certificates
            *   Logs appended to (not overwrite) the /var/log/shim.log
            *   Logs atomically wiped when 512kb reached
            *   Installer now has color! and logo


... time unknown ...    Shim born (as part of the CloudSNR monitoring project)




installer.sh
#! /bin/sh

# Userify Shim Installer
# Copyright (c) 2011-2015 Userify Corporation

# How the shim works:
#
# 1. Installer creates /opt/userify/ containing:
#
#       a.  /opt/userify/creds.py
#       b.  /opt/userify/shim.sh - should start on reboot
#       c.  /opt/userify/uninstall.sh
#           
#       d.  /var/log/shim.log - check this for output
#
#    Review shim.sh and uninstall.sh in this file below
#    Entire /opt/userify directory and /var/log/shim.log is root-only.
#
#
# 2. shim.sh, which should start on reboot, loops endlessly:
#
#       a.  download and run shim.py.
#
#       b.  shim.py reads creds.py and initiates a request for
#           the latest user list.
#
#       c.  upon making needed changes, shim.py delays before
#           exiting.
#

export RED_TEXT="[31m"
export BLUE_TEXT="[34m"
export GREEN_TEXT="[32m"
export PURPLE_TEXT="[35m"
export CYAN_TEXT="[36m"
export RESET_TEXT="[0m"

clear
cat << EOF
   
             ${BLUE_TEXT}            _--_
             ${BLUE_TEXT}           (    \\
             ${BLUE_TEXT}        --/      )
             ${BLUE_TEXT}   .-- /   \\      \\
             ${BLUE_TEXT} ./   \\            )${PURPLE_TEXT} _  __
             ${BLUE_TEXT}/${GREEN_TEXT}_   _ ___  ___ _ __${PURPLE_TEXT}(_)/ _|_   _
             ${GREEN_TEXT}| | | / __|/ _ \ '__${PURPLE_TEXT}| | |_  | | |
             ${GREEN_TEXT}| |_| \__ \  __/ |  ${PURPLE_TEXT}| |  _| |_| |
             ${GREEN_TEXT} \__,_|___/\___|_|  ${PURPLE_TEXT}|_|_|  \__, |
             ${GREEN_TEXT}                    ${PURPLE_TEXT}       |___/  ${GREEN_TEXT}tm
${RESET_TEXT}

[37;42m                Installing Userify now..                     ${RESET_TEXT}
-------------------------------------------------------------
${PURPLE_TEXT}Tip: to understand how the shim works, read the source at
${CYAN_TEXT}https://github.com/userify/shim/
${RESET_TEXT}
EOF

# Check for root
if [ "$(id -u)" != "0" ]; then
    cat << EOF >&2
${RED_TEXT}
Unfortunately, the Userify Shim requires root permissions in order to
create user accounts and manage sudoers. Please review the shim
source code at https://github.com/userify/shim
${RESET_TEXT}
EOF
    exit 1
fi


# Check for Linux
if [ "$(uname -s)" != "Linux" ]; then
    echo "${RED_TEXT}Currently, Userify supports only Linux systems.${RESET_TEXT}" >&2
fi


# Attempt to install sudo on Debian
# (might not be included on some very minimal/netinst Debian systems)
# set +e
# apt-get update 2>/dev/null
# apt-get -y install sudo 2>/dev/null
# set -e


set -e
[ -d /opt/userify ] && (
    echo "${RED_TEXT}Please remove /opt/userify, or execute
    ${GREEN_TEXT} sudo ${BLUE_TEXT}/opt/userify/uninstall.sh
${RED_TEXT}before continuing.${RESET_TEXT}" >&2
    exit -1
)


echo "${GREEN_TEXT}Creating Userify directory (/opt/userify/)${RESET_TEXT}"
mkdir /opt/userify/ || (
    echo "${RED_TEXT}Unable to create directory /opt/userify.${RESET_TEXT}" >&2
    exit 1
)


# Create uninstall.sh script in /opt/userify

echo "${GREEN_TEXT}Creating uninstall script (/opt/userify/uninstall.sh)${RESET_TEXT}"
cat << EOF > /opt/userify/uninstall.sh
#! /bin/sh +e

# --------------------------------------------
#
# uninstall.sh
# This script uninstalls the entire Userify agent and kills
# off any shim processes that are still running.
#
# --------------------------------------------

# Copyright (c) 2015 Userify Corp.

echo
echo
echo -------------------------------------------------------------
echo "[31mRemoving Userify...[0m"

# Debian, Ubuntu, RHEL:
sed -i "s/\/opt\/userify\/shim.sh \&//" \
    /etc/rc.local 2>/dev/null

# SUSE:
sed -i "s/\/opt\/userify\/shim.sh \&//" \
    /etc/init.d/after.local 2>/dev/null

# Fedora:
# systemctl disable userify-shim.service 2>/dev/null
# rm -f /etc/systemd/system/userify-shim.service 2>/dev/null

# Wipe out entire /opt/userify directory
rm -Rf /opt/userify/

# Kill off remaining shim processes
killall shim.py shim.sh 2>/dev/null

echo [32m

echo Finished!
echo Userify has been removed, but the user accounts it created still
echo exist and no changes have been made to them.
echo You can now install a new Userify shim as desired.
echo [0m-------------------------------------------------------------
echo
echo

EOF



if [ "x$api_id" != "x" ]; then
    echo "${GREEN_TEXT}Creating API login config (/opt/userify/creds.py)${RESET_TEXT}"
    echo -n > /opt/userify/creds.py
    chmod 0600 /opt/userify/creds.py
    # create creds configuration file
    cat <<EOF >> /opt/userify/creds.py
# Move this server to a different server group
# by editing these.

api_id = "$api_id"
api_key = "$api_key"

EOF

    cat <<EOF >> /opt/userify/userify_config.py
# Enable this to receive additional verbosity in /var/log/shim.log
debug = False

# Enable this to not actually make changes.
# This can also be used to temporary disable the shim.
dry_run = False

# Changing these requires appropriate licensing.
shim_host = "$shim_host"
self_signed = $self_signed

EOF

else
    echo "${RED_TEXT}api_id variable not found, skipping creds.py creation."
    echo "This might be a bug unless you did this on purpose."
    echo "NOTE, Userify cannot work without creds.py. Please create it yourself."
    echo ${RESET_TEXT}
fi


# Create shim.sh script in /opt/userify

echo "${GREEN_TEXT}Creating shim (/opt/userify/shim.{sh,py})${RESET_TEXT}"
cat << "EOF" > /opt/userify/shim.sh
#! /bin/bash +e

# --------------------------------------------
#
# shim.sh
# This is the script that actually calls
# shim.py.
#
# --------------------------------------------

# Copyright (c) 2015 Userify Corp.

# keep shim.log from getting too big
touch /var/log/shim.log
[[ $(find /var/log/shim.log -type f -size +524288c 2>/dev/null) ]] && \
    rm /var/log/shim.log
touch /var/log/shim.log
chmod -R 600 /var/log/shim.log


# kick off shim.py
[ -z "$PYTHON" ] && PYTHON="$(which python)"
output=$(curl -kSs https://shim.userify.com/shim.py | $PYTHON 2>&1)
echo "$output" >> /var/log/shim.log


# a little extra fix for thundering herd
sleep $(( ( RANDOM % 5 )  + 1 ))


# call myself. fork before exiting.
/opt/userify/shim.sh &

EOF


echo "${GREEN_TEXT}Removing exit 0 from rc.local (if there)${RESET_TEXT}"
set +e
sed -i "s/^ *exit 0.*/# &/" /etc/rc.local 2>/dev/null
set -e

echo "${GREEN_TEXT}Checking Shim Startup${RESET_TEXT}"

# most Linux versions can manage with a line added to rc.local:
if [ -f /etc/rc.local ]; then
    distro="Linux"
    fname=/etc/rc.local
elif [ -f /etc/init.d/after.local ]; then
    distro="SUSE"
    fname=/etc/init.d/after.local
# elif [ -f /etc/fedora-release ]; then
#     distro="Fedora"
#     cat << EOF > /etc/systemd/system/userify-shim.service
# [Unit]
# Description=Userify Shim (userify.com)
# 
# [Service]
# Type=forking
# ExecStart=/opt/userify/shim.sh
# 
# [Install]
# WantedBy=multi-user.target
# EOF
#     systemctl enable userify-shim.service
else
    cat << EOF >&2
${RED_TEXT}
Unable to set start at bootup -- no /etc/rc.local file?
You'll have to set shim to startup on it's own: create an
init script that launches /opt/userify/shim.sh on startup.
In most distributions, this would have been a single line
in /etc/rc.local, but you may need to do something more
exotic. Please contact us with Linux version information
and we may have more information for you.${RESET_TEXT}
EOF
    exit 1
fi


# set +e
# # RHEL7/Fedora hack:
# if [ -f /etc/rc.d/rc.local ]; then
#     set +e; mv /etc/rc.local /etc/rc.local.old; set -e
#     ln -s /etc/rc.d/rc.local /etc/rc.local
# fi
# set -e


# actually set up the startup
if [ "$distro" != "Fedora" ]; then
    # remove any existing lines:
    set +e
        sed -i "s/\/opt\/userify\/shim.sh \&//" "$fname" 2>/dev/null
    set -e
    echo "${GREEN_TEXT}Adding $distro Startup Script to $fname${RESET_TEXT}"
    echo >> "$fname"
    echo "/opt/userify/shim.sh &" >> "$fname"
fi

echo "${GREEN_TEXT}Setting Permissions${RESET_TEXT}"
chmod -R 700 \
    /opt/userify/ \
    /opt/userify/uninstall.sh \
    /opt/userify/shim.sh
[ -f /var/log/shim.log ] && rm /var/log/shim.log
touch /var/log/shim.log
set +e
chmod +x /etc/rc.local 2>/dev/null
# RHEL7:
chmod +x /etc/rc.d/rc.local 2>/dev/null


echo "${GREEN_TEXT}Launching shim.sh${RESET_TEXT}"
set +e;
killall shim.py shim.sh 2>/dev/null
set -e
/opt/userify/shim.sh &

echo
echo "${PURPLE_TEXT}Finished. Userify shim has been completely installed."
echo "/opt/userify/uninstall.sh as root to uninstall."
echo "Please review first shim output in /var/log/shim.log."
# echo "(wait a few seconds..)"
# echo ${BLUE_TEXT}
# sleep 2
# output=$(cat /var/log/shim.log)
# if [ "x$output" == "x" ]; then
#     echo ${RED_TEXT}
#     echo Unable to review shim.log, please review it separately
#     echo to ensure the shim is working properly.
    echo i.e.,:  ${BLUE_TEXT}cat /var/log/shim.log${RESET_TEXT}
# else
#     echo $OUTPUT
# fi
echo ${GREEN_TEXT}
echo "Thanks for using Userify!"
echo ${RESET_TEXT}
echo -------------------------------------------------------------
echo




shim.py
#! /usr/bin/env python

# Userify Shim
# Copyright (c) 2012-2014 Userify

# Note: the shim can be easily installed and configured via the shim installer,
# or just installed manually after creating the creds.py file.

# base level imports

try:
    import json
except:
    import simplejson as json

# Standard Library Imports
import subprocess
import os
import os.path
import signal
import httplib
import sys
import datetime
import time
import traceback
import base64
import urllib
import random
from pprint import pprint
# catch stderr
from subprocess import PIPE as pipe

sys.path.append("/opt/userify")
import creds


try:
    import userify_config as config
except:
    import creds as config


self_signed = getattr(config, "self_signed", False)
dry_run = getattr(config, "dry_run", False)
shim_host = getattr(config, "shim_host", "configure.userify.com")
debug = getattr(config, "debug", False)


# check for self_signed
ssl_security_context = None
if self_signed:
    try:
        # fails on python < 2.6:
        import ssl
        # not avail in python < 2.7:
        ssl_security_context = (hasattr(ssl, '_create_unverified_context') 
            and ssl._create_unverified_context() or None)
    except:
        print "Self signed access attempted, but unable to open self-signed"
        print "security context. This Python may not support (or need) that."
        raise


# local_download = urllib.urlretrieve

def userdel(username, permanent=False):
    # removes user and renames homedir
    removed_dir = "/home/deleted:" + username
    home_dir = "/home/" + username
    if not permanent:
        if os.path.isdir(removed_dir):
            qexec(["/bin/rm", "-Rf", removed_dir])
        # Debian:
        qexec(["/usr/bin/pkill", "--signal", "9", "-u", username])
        # RHEL:
        qexec(["/usr/bin/pkill", "-9", "-u", username])
        qexec(["/usr/sbin/userdel", username])
        qexec(["/bin/mv", home_dir, removed_dir])
    else:
        qexec(["/usr/sbin/userdel", "-r", username])


def useradd(name, username, preferred_shell):
    removed_dir = "/home/deleted:" + username
    home_dir = "/home/" + username

    if dry_run:
        print "DRY RUN: Adding user ", name, username, preferred_shell
        return

    # restore removed home directory
    if not os.path.isdir(home_dir) and os.path.isdir(removed_dir):
        qexec(["/bin/mv", removed_dir, home_dir])
    if os.path.isdir(home_dir):
        useradd_suffix = ""
    else:
        useradd_suffix = "-m"
    cmd = ["/usr/sbin/useradd", useradd_suffix,
        "--comment", "userify-" + name,
        "-s", preferred_shell if preferred_shell else "/bin/bash",
        "--user-group", username]
    subprocess.call([i for i in cmd if i])
    fullchown(username, home_dir)
    parse_passwd()


def sudoers_add(username, perm=""):
    fname = "/etc/sudoers.d/" + username
    text = sudoerstext(username, perm)

    if dry_run:
        print "DRY RUN: Adding sudoers: ", fname, text
        return

    if perm:
        if not os.path.isfile(fname) or open(fname).read() != text:
            open(fname, "w").write(text)
            fullchmod("0440", fname)
    else:
        sudoers_del(username)


def sudoers_del(username):
    fname = "/etc/sudoers.d/" + username
    if os.path.isfile(fname):
        qexec(["/bin/rm", fname])


def sudoerstext(username, perm):
    return "\n".join((
        "# Generated by userify",
        username + " "*10 + perm, ""))


def sshkeytext(ssh_public_key):
    return "\n".join((
        "# Generated by userify",
        "# Paste your new key at console.userify.com.", ssh_public_key, ""))


def sshkey_add(username, ssh_public_key=""):

    if not ssh_public_key:
        return

    userpath = "/home/" + username
    sshpath = userpath + "/.ssh/"

    if dry_run:
        print "DRY RUN: Adding user ssh key", sshpath, ssh_public_key
        return

    failsafe_mkdir(sshpath)
    fname = sshpath + "authorized_keys"
    text = sshkeytext(ssh_public_key)
    if not os.path.isfile(fname) or open(fname).read() != text:
        open(fname, "w").write(text)
        fullchown(username, sshpath)


def fullchown(username, path):
    qexec(["chown", "-R", username+":"+username, path])


def fullchmod(mode, path):
    qexec(["chmod", "-R", mode, path])


def qexec(cmd):
    print "[shim] exec: \"" + " ".join(cmd) + '"'
    try: subprocess.check_call(cmd)
         # ,stderr=pipe, stdout=pipe)
    except:
        traceback.print_exc()
        pass


def failsafe_mkdir(path):
    try: os.mkdir(path)
    except OSError: pass


def auth(id,key):
    return base64.b64encode(":".join((creds.api_id, creds.api_key)))


def https(method, path, data=""):

    # timeout is crucial to prevent eternal hangs, but
    # wasn't available until Python 2.5 (so shim won't
    # work on ancient versions of RHEL <= 5)

    if ssl_security_context:
        h = httplib.HTTPSConnection(shim_host, timeout=30,
                context=ssl_security_context)
    else:
        h = httplib.HTTPSConnection(shim_host, timeout=30)

    headers = {
        "Accept": "text/plain, */json",
        "Authorization": "Basic " + auth(creds.api_id, creds.api_key)
    }
    h.request(method, path, data, headers)
    return h


def parse_passwd():
    # returns a list of passwd lines, ordered as
    # username, unused, uid, gid, comment, homedir, shell
    app["passwd"] = [[i.strip() for i in l.split(":")]
        for l in open("/etc/passwd").read().strip().split("\n")]
    app["passwd"] = [i if len(i)>6 else i.append("") for i in app["passwd"]]


def current_usernames():
    return [user[0] for user in app["passwd"]]


def current_userify_users():
    "get only usernames created by userify"
    return [user for user in app["passwd"] if user[4].startswith("userify-")]


def remove_user(username, permanent=False):
    # completely removes user
    if dry_run:
        print "DRY RUN: Removing user: ", username, "permanently:", permanent
        return
    try: userdel(username, permanent)
    except: pass
    try: sudoers_del(username)
    except: pass


def process_users(good_users):
    for username, user in good_users.iteritems():
        if username not in current_usernames():
            useradd(user["name"], username, user["preferred_shell"])
        if "ssh_public_key" in user:
            sshkey_add(username, user["ssh_public_key"])
        sudoers_add(username, user["perm"])
    for userrow in current_userify_users():
        username = userrow[0]
        if username not in good_users.keys():
            print "[shim] removing" + username
            remove_user(username)


def main():
    parse_passwd()
    h = https("POST", "/api/userify/configure")
    response = h.getresponse()
    text = response.read()
    failure = response.status != 200
    if debug or failure:
        print response.status, response.reason
        pprint(text)
    try:
        configuration = json.loads(text)
        if debug or failure:
            pprint(configuration)
        if failure and "error" in configuration:
            print "\n", response.reason.upper(), configuration["error"]
    except:
        failure = True
    if "error" in configuration or failure:
        return 3
    process_users(configuration["users"])
    return configuration["shim-delay"] if "shim-delay" in configuration else 1



app = {}
if __name__ == "__main__":
    try:
        print
        print '-'*30
        print "[shim] start: %s" % time.ctime()
        s = time.time()
        time_to_wait = main()
        elapsed = time.time() - s
        print "[shim] elapsed: " + str(int(elapsed * 1000)/1000.0) + "s"
        if elapsed < time_to_wait:
            print "[shim] sleeping: %s" % (time_to_wait-elapsed)
            time.sleep(time_to_wait-elapsed)
    except:
        time.sleep(3)
        # display error to stdout
        # and attempt restart via shim.sh
        raise
    print '-'*30
    print
